File: pruebaAppBar/app/src/main/java/com/p2/pruebaappbar/FragmentPagerAdapter.java
package com.p2.pruebaappbar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import java.util.ArrayList;
import java.util.Arrays;

public class FragmentPagerAdapter extends androidx.fragment.app.FragmentStatePagerAdapter  {

    /* Este Array representa los títulos de las pestañas */
    ArrayList<String> misTabs = new ArrayList<>(Arrays.asList("Suburbia","NCTH","TGG","Sister cities"
    ,"Disco Creado","Disco CAMBIADO!!"));

    /*Lista de los discos que se están mostrando */
    ArrayList<Disco> listaDiscos;

    public ArrayList<Disco> getListaDiscos() {
        return listaDiscos;
    }

    public void setListaDiscos(ArrayList<Disco> listaDiscos) {
        this.listaDiscos = listaDiscos;
    }

    /*Constructor por defecto:
    * fm: El objeto Java que se encarga de gestionar los Fragment: crearlos, borrarlos, pararlos...
    * behavior: Puede ser BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT o BEHAVIOR_SET_USER_VISIBLE_HINT */
    public FragmentPagerAdapter(@NonNull FragmentManager fm, int behavior) {
        super(fm, behavior);
    }


    @NonNull
    @Override
    public Fragment getItem(int position) {
        /*Devolvemos el fragmento que esté en la posición position.
        Para ello creamos un nuevo Fragmento y le pasamos la lista de los discos total y el filtro por el que
        discernir
         */
        return new Fragmento(listaDiscos,getPageTitle(position).toString());
    }

    @Override
    public int getCount() {
        return misTabs.size();
    }

    @Nullable
    @Override
    public CharSequence getPageTitle(int position) {
        return misTabs.get(position);
    }

    /*Forma un poco más compleja de sobreescribir el objeto:
    Lo que se hace es modificar el fragmento que se ha modificado, es decir, el que se ha notificado que ha
    habido un cambio, y actualizar la lista de discos.
    Véase la función updateRecyclerView de la clase Fragmento, que lo que hace es regenerar la lista
    en el recyclerView.
     */
    @Override
    public int getItemPosition(@NonNull Object object) {
        if(object instanceof Fragmento) ((Fragmento) object).updateRecyclerView(listaDiscos);
        return super.getItemPosition(object);
    }

    /*
    Otra forma más sencilla de implementar el evento
    @Override
    public int getItemPosition(@NonNull Object object) {
        return POSITION_NONE;
    }

     */
}

File: pruebaAppBar/app/src/main/java/com/p2/pruebaappbar/FragmentPagerAdapter.kt
package com.p2.pruebaappbar


import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.adapter.FragmentViewHolder
import java.util.*
import kotlin.collections.ArrayList

/*Constructor por defecto:
    * fm: Una subclase de Activity que representa un fragmento
*/
class FragmentPagerAdapter
(var fm: FragmentActivity) : FragmentStateAdapter(fm){
    /* Este Array representa los títulos de las pestañas */
    var misTabs = ArrayList(Arrays.asList("Suburbia", "NCTH", "TGG", "Sister cities", "Disco Creado", "Disco CAMBIADO!!"))

    /*Lista de los discos que se están mostrando */
    var listaDiscos: ArrayList<Disco> = ArrayList()

    /*Devolvemos el fragmento que esté en la posición position.
      Para ello creamos un nuevo Fragmento y le pasamos la lista de los discos total y el filtro por el que
      discernir.
      Lo normal será que tengamos que usar una estructura condicional para elegir que fragmento
      hay que renderizar.
   */
    override fun createFragment(position: Int): Fragment {
        return Fragmento(listaDiscos, getPageTitle(position).toString())
    }

    /*
    * Esta función devuelve el numero de items (pestañas)
    */
    override fun getItemCount(): Int {
        return misTabs.size
    }


    // Esta función devuelve el nombre de la pestaña asociada a positon.
    fun getPageTitle(position: Int): CharSequence? {
        return misTabs[position]
    }

    /* Este método es similar al que hay implementado en el RecyclerView, aquí se llamará cada vez
     * que se llame a algún método de notificación.
     * Recibe la pestaña pulsada (position) y los datos que se han modificado (payloads).
     * Cada pestaña, internamente, se identifica con f seguido de un número, representando la posición
     * de la pestaña -> f0,f1..
     */
    override fun onBindViewHolder(
            holder: FragmentViewHolder,
            position: Int,
            payloads: MutableList<Any>
    ) {
        val tag = "f$position"
        // Se obtiene el fragmento que está siendo mostrado actualmente
        val fragment = fm.supportFragmentManager.findFragmentByTag(tag)
        if (fragment != null) {
                // Se obtiene y se refresca manualmente.
                (fragment as Fragmento).actualizarLista(payloads[0] as ArrayList<Disco>)
        } else {
                super.onBindViewHolder(holder, position, payloads)
        }
    }

    /* Siguiendo una estructura similar al método anterior, obtendo el fragmento actual, y se llama
     * al método borrarElemento()
     */
    fun borrarElemento(pestanya: Int){
        val tag = "f$pestanya"
        val fragment = fm.supportFragmentManager.findFragmentByTag(tag)
        (fragment as Fragmento).borrarElemento()

    }

}